### ISSUE TITLE
Question about kpt_shape parameter and handling missing keypoints in YOLOv8 pose estimation

### ISSUE THREAD

**Issue by jindameias** [ultralytics/ultralytics#17116]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

# Description:

Hello, I am encountering some confusion while using Ultralytics YOLOv8 for pose estimation.

# Issue Description:

I am working on a project where there are a total of 4 keypoints, but not every object has all 4 keypoints. In other words, some objects may have missing keypoints. I am confused about how to configure the kpt_shape parameter and how to handle these missing keypoints.

# Specific Questions:

In the configuration file, I have set kpt_shape to [4, 2], meaning 4 keypoints with (x, y) coordinates. Is this configuration correct, or is there a better way to specify it?

For the training data, if some objects are missing keypoints, how should these missing keypoints be annotated? My current idea is to set them to (0, 0) or (-1, -1). Which method is recommended, or do you have any other suggestions?

During the inference stage, if the result contains keypoints with zero confidence (since we are not using the confidence value directly), should these keypoints be filtered out during post-processing?

# Example Configuration File:

**yaml**

```
# Model configuration file
nc: 1  # Number of classes
kpt_shape: [4, 2]  # 4 keypoints, each with (x, y) coordinates
```

# Example Inference Code:

```

from ultralytics import YOLO

# Load model
model = YOLO('train_model.pt')

# Inference
results = model('test_image.jpg')

# Process results
for pose in results.keypoints:
    keypoints = pose.xyxy.numpy()  # Assuming `xyxy` contains (x, y) data
    for idx, kpt in enumerate(keypoints):
        x, y = kpt
        if (x, y) != (0, 0) and (x, y) != (-1, -1):  # Filter out missing keypoints
            print(f"Keypoint {idx}: (x, y)=({x}, {y})")
```
I really appreciate the efforts of the developer team and thank you for your assistance!

# Additional Information:

If you need more information or sample data, please let me know.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @jindameias, thank you for your interest in Ultralytics 🚀! Your question about handling missing keypoints and configuring the `kpt_shape` parameter in your YOLOv8 pose estimation project is indeed an important one.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us debug it. 

For your ❓ Question on custom training and keypoint handling, it's great that you've shared your current configuration and inference examples. Please feel free to provide additional details such as dataset examples or training logs, which could be helpful in addressing your queries more effectively. Additionally, you might want to check our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to ensure optimal configurations.

Your approach to filtering out zero-confidence keypoints during post-processing is a practical one. The community and our documentation could have further insights on best practices for such scenarios, and you might benefit from engaging with other experts on our [Discourse](https://community.ultralytics.com) or [Discord](https://ultralytics.com/discord) 🎧 for real-time support.

## Upgrade

To make sure you're using the latest features and fixes, upgrade your `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

Test your code in one of our verified environments which come with all necessary dependencies preinstalled:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud**, **Amazon** Deep Learning AMIs, and Docker images are available with guides for setup in our documentation.

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing, ensuring that issues are less likely to stem from a bug in the system.

This is an automated response 📝. An Ultralytics engineer will also review your issue soon to provide additional assistance. Thank you for your patience and for contributing to our community!

**Comment by Y-T-G** [comment]
You should use a shape of `[4,3]`. The third value would be visibility which will be set to 0 along with the coordinates for keypoints that don't exist. And the visibility is 1 if they exist.

**Comment by Y-T-G** [comment]
Correct
